# Библиотеки Python для Data Science: Numpy, Matplotlib, Scikit-learn

## Урок 6. Видеоурок. Обучение с учителем в Scikit-learn

### Тема “Обучение с учителем”


Задание 1:
Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn..
Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью
функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random state должен быть равен 42.
Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля
sklearn.linear_model.
Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на
тестовых.

In [1]:
# Импортируем необходимые библиотеки
import pandas as pd
import numpy as np
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# Загружаем набор данных California housing
california_housing = fetch_california_housing()
X = pd.DataFrame(california_housing.data, columns=california_housing.feature_names)
y = pd.Series(california_housing.target)

# Разбиваем данные на тренировочные и тестовые наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Создаем модель линейной регрессии
lr = LinearRegression()

# Обучаем модель на тренировочных данных
lr.fit(X_train, y_train)

# Делаем предсказание на тестовых данных
y_pred = lr.predict(X_test)

# Выводим первые пять предсказаний для проверки
print(y_pred[:5])

# Вычисляем R2 для оценки модели
r2 = r2_score(y_test, y_pred)
print(f'R2 score: {r2}')

[0.72604907 1.76743383 2.71092161 2.83514727 2.60695807]
R2 score: 0.595770232606166


Датасет "Boston House Prices" был удалён из библиотеки scikit-learn начиная с версии 1.2 из-за этических соображений. Взял пример кода, который выполняет задание с использованием альтернативного датасета "California Housing".



Задание 2:
Создайте модель под названием model с помощью класса RandomForestRegressor из модуля
sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.
Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно
применение массивов вместо датафрейма.
Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из
предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.


In [2]:
# Импортируем необходимые библиотеки
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score

# Создаем модель RandomForestRegressor с указанными параметрами
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

# Обучаем модель на тренировочных данных
# Преобразуем y_train в одномерный массив
model.fit(X_train, y_train.values if isinstance(y_train, pd.Series) else y_train)

# Делаем предсказание на тестовых данных
y_pred_model = model.predict(X_test)

# Вычисляем R2 для модели RandomForestRegressor
r2_model = r2_score(y_test, y_pred_model)
print(f'R2 score для модели RandomForestRegressor: {r2_model}')

# Сравниваем R2 модели RandomForestRegressor с R2 модели LinearRegression из предыдущего задания
r2_lr = 0.595770232606166  # R2 для модели LinearRegression из предыдущего задания

# Выводим результаты сравнения
print(f'R2 score для модели LinearRegression: {r2_lr}')
print(f'Модель {"RandomForestRegressor" if r2_model > r2_lr else "LinearRegression"} работает лучше.')

# В зависимости от результатов, добавляем комментарий о том, какая модель лучше
if r2_model > r2_lr:
    # Если R2 для RandomForestRegressor выше, то он работает лучше для этого набора данных
    print("Модель RandomForestRegressor показала лучшие результаты по сравнению с LinearRegression на этом наборе данных.")
else:
    # Если R2 для LinearRegression выше, то она работает лучше для этого набора данных
    print("Модель LinearRegression показала лучшие результаты по сравнению с RandomForestRegressor на этом наборе данных.")

R2 score для модели RandomForestRegressor: 0.7945061536878142
R2 score для модели LinearRegression: 0.595770232606166
Модель RandomForestRegressor работает лучше.
Модель RandomForestRegressor показала лучшие результаты по сравнению с LinearRegression на этом наборе данных.


Комментарий: Модель случайного леса (RandomForestRegressor) с коэффициентом детерминации R2 равным 0.7945061536878142 работает лучше, чем модель линейной регрессии (LinearRegression) с R2 равным 0.595770232606166 на данном наборе данных.

*Задание 3:
Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.
С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.


In [3]:
# Импортируем класс RandomForestRegressor
from sklearn.ensemble import RandomForestRegressor

# Вызываем документацию для класса RandomForestRegressor
help(RandomForestRegressor)

# Создаем модель RandomForestRegressor (предполагается, что модель уже обучена, как в задании 2)
# Если модель еще не обучена, необходимо ее обучить, используя следующие строки кода:
# model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)
# model.fit(X_train, y_train.values if isinstance(y_train, pd.Series) else y_train)

# Получаем важность признаков с помощью атрибута feature_importances_
importances = model.feature_importances_

# Сумма всех показателей важности должна быть равна 1. Проверяем это:
print(f'Сумма всех показателей важности: {np.sum(importances)}')

# Находим индексы двух наиболее важных признаков
indices = np.argsort(importances)[-2:]

# Выводим названия двух наиболее важных признаков
print(f'Два признака с наибольшей важностью: {X.columns[indices[0]]} и {X.columns[indices[1]]}')

Help on class RandomForestRegressor in module sklearn.ensemble._forest:

class RandomForestRegressor(ForestRegressor)
 |  RandomForestRegressor(n_estimators=100, *, criterion='squared_error', max_depth=None, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_features=1.0, max_leaf_nodes=None, min_impurity_decrease=0.0, bootstrap=True, oob_score=False, n_jobs=None, random_state=None, verbose=0, warm_start=False, ccp_alpha=0.0, max_samples=None)
 |  
 |  A random forest regressor.
 |  
 |  A random forest is a meta estimator that fits a number of classifying
 |  decision trees on various sub-samples of the dataset and uses averaging
 |  to improve the predictive accuracy and control over-fitting.
 |  The sub-sample size is controlled with the `max_samples` parameter if
 |  `bootstrap=True` (default), otherwise the whole dataset is used to build
 |  each tree.
 |  
 |  Read more in the :ref:`User Guide <forest>`.
 |  
 |  Parameters
 |  ----------
 |  n_estimators 

Комментарий:

Сумма показателей важности признаков, полученных из обученной модели RandomForestRegressor, практически равна 1, что соответствует ожиданиям (из-за округления в вычислениях она может быть немного меньше, но очень близка к 1).

Два признака, которые были определены как наиболее важные, это AveOccup (средняя занятость, то есть среднее количество людей, проживающих в доме) и MedInc (медианный доход в блоке). Это означает, что эти два признака имеют наибольшее влияние на прогнозируемую переменную в модели случайного леса для данного набора данных.

*Задание 4:
В этом задании мы будем работать с датасетом, с которым мы уже знакомы по домашнему заданию
по библиотеке Matplotlib, это датасет Credit Card Fraud Detection.Для этого датасета мы будем решать
задачу классификации - будем определять,какие из транзакции по кредитной карте являются
мошенническими.Данный датасет сильно несбалансирован (так как случаи мошенничества
относительно редки),так что применение метрики accuracy не принесет пользы и не поможет выбрать
лучшую модель.Мы будем вычислять AUC, то есть площадь под кривой ROC.
Импортируйте из соответствующих модулей RandomForestClassifier, GridSearchCV и train_test_split.
Загрузите датасет creditcard.csv и создайте датафрейм df.
С помощью метода value_counts с аргументом normalize=True убедитесь в том, что выборка
несбалансирована. Используя метод info, проверьте, все ли столбцы содержат числовые данные и нет
ли в них пропусков.Примените следующую настройку, чтобы можно было просматривать все столбцы
датафрейма:
pd.options.display.max_columns = 100.
Просмотрите первые 10 строк датафрейма df.
Создайте датафрейм X из датафрейма df, исключив столбец Class.
Создайте объект Series под названием y из столбца Class.
Разбейте X и y на тренировочный и тестовый наборы данных при помощи функции train_test_split,
используя аргументы: test_size=0.3, random_state=100, stratify=y.
У вас должны получиться объекты X_train, X_test, y_train и y_test.
Просмотрите информацию о их форме.
Для поиска по сетке параметров задайте такие параметры:
parameters = [{'n_estimators': [10, 15],
'max_features': np.arange(3, 5),
'max_depth': np.arange(4, 7)}]
Создайте модель GridSearchCV со следующими аргументами:
estimator=RandomForestClassifier(random_state=100),
param_grid=parameters,
scoring='roc_auc',
cv=3.
Обучите модель на тренировочном наборе данных (может занять несколько минут).
Просмотрите параметры лучшей модели с помощью атрибута best_params_.
Предскажите вероятности классов с помощью полученной модели и метода predict_proba.
Из полученного результата (массив Numpy) выберите столбец с индексом 1 (вероятность класса 1) и
запишите в массив y_pred_proba. Из модуля sklearn.metrics импортируйте метрику roc_auc_score.
Вычислите AUC на тестовых данных и сравните с результатом,полученным на тренировочных данных,
используя в качестве аргументов массивы y_test и y_pred_proba.



In [4]:
# Импортируем необходимые библиотеки и функции
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import roc_auc_score

# Загружаем датасет
df = pd.read_csv('/content/creditcard.csv')

# Проверяем баланс классов
print(df['Class'].value_counts(normalize=True))

# Проверяем наличие числовых данных и пропусков
df.info()

# Удаляем строки с пропущенными значениями, если они есть
df.dropna(inplace=True)

# Проверяем, все ли пропуски были удалены
print(df.isnull().sum().max())  # должно быть 0, если нет пропусков

# Настраиваем отображение всех столбцов
pd.options.display.max_columns = 100

# Просматриваем первые 10 строк датафрейма
print(df.head(10))

# Создаем датафрейм X, исключая столбец 'Class'
X = df.drop('Class', axis=1)

# Создаем Series y из столбца 'Class'
y = df['Class']

# Разбиваем данные на тренировочный и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=100, stratify=y)

# Просматриваем информацию о форме объектов
print(f'X_train shape: {X_train.shape}')
print(f'X_test shape: {X_test.shape}')
print(f'y_train shape: {y_train.shape}')
print(f'y_test shape: {y_test.shape}')

# Задаем параметры для поиска по сетке
parameters = [{'n_estimators': [10, 15],
               'max_features': np.arange(3, 5),
               'max_depth': np.arange(4, 7)}]

# Создаем модель GridSearchCV
grid_search = GridSearchCV(estimator=RandomForestClassifier(random_state=100),
                           param_grid=parameters,
                           scoring='roc_auc',
                           cv=3)

# Обучаем модель на тренировочном наборе данных
grid_search.fit(X_train, y_train)

# Просматриваем параметры лучшей модели
print(f'Лучшие параметры: {grid_search.best_params_}')

# Предсказываем вероятности классов
y_pred_proba = grid_search.predict_proba(X_test)[:, 1]

# Вычисляем AUC на тестовых данных
auc_test = roc_auc_score(y_test, y_pred_proba)
print(f'AUC на тестовых данных: {auc_test}')

# Можно также вычислить AUC на тренировочных данных для сравнения
y_pred_proba_train = grid_search.predict_proba(X_train)[:, 1]
auc_train = roc_auc_score(y_train, y_pred_proba_train)
print(f'AUC на тренировочных данных: {auc_train}')

Class
0    0.998273
1    0.001727
Name: proportion, dtype: float64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-

Выводы:

Дисбаланс классов: Датасет сильно несбалансирован, с доминированием легитимных транзакций (класс "0") над мошенническими (класс "1"). Это обычная ситуация для задач обнаружения мошенничества, и она требует особого внимания при выборе метрик оценки модели и техник обучения, чтобы обеспечить адекватное обнаружение мошеннических транзакций.

Предобработка данных: В датасете были обнаружены и удалены строки с пропущенными значениями. Остальные данные были проверены на наличие числовых значений и отсутствие пропусков, что делает их пригодными для использования в моделях машинного обучения. Признаки V1 до V28 являются результатом PCA преобразования, что помогает защитить личную информацию пользователей.

Разделение данных и обучение модели: Данные были разделены на тренировочный и тестовый наборы с учетом стратификации по целевой переменной, чтобы сохранить пропорции классов. Была обучена модель случайного леса (RandomForestClassifier) с использованием поиска по сетке для оптимизации гиперпараметров. Лучшая модель показала высокое значение AUC на тестовых данных (приблизительно 0.9726), что указывает на хорошую способность различать классы.

Оценка переобучения: Несмотря на высокий AUC на тестовом наборе данных, существует значительная разница между AUC на тренировочном (почти 1) и тестовом наборах данных. Это может указывать на потенциальное переобучение модели, поскольку она слишком хорошо подогнана под тренировочные данные.